pkg/composition/controller2.go

```diff
@@ -0,0 +1,624 @@
+	for _, fo := range fOr {
+		result = append(result, *fo.namespace)
+	}
+	return result
+}
+
+func (c *Controller) Process2(ctx *ctrl.ProcessContext) (bool /* retriable */, error) {
+	sd := ctx.Object.(*comp_v1.ServiceDescriptor)
+	serviceName := sd.Name
+	logger := ctx.Logger.With(zap.String("sd_name", sd.Name))
+	logger.Debug("Started processing ServiceDescriptor")
+
+	// If SD is marked for deletion and SD finalizer is missing, we have nothing to do
+	if sd.ObjectMeta.DeletionTimestamp != nil && !hasServiceDescriptorFinalizer(sd) {
+		logger.Sugar().Infof("No SD finalizer for %q. Skipping", sd.Name)
+		return false, nil
+	}
+
+	// If SD finalizer is missing, add it and finish the processing iteration
+	if sd.ObjectMeta.DeletionTimestamp == nil && !hasServiceDescriptorFinalizer(sd) {
+		sd.Finalizers = addServiceDescriptorFinalizer(sd.GetFinalizers())
+		_, retriableStatus, errStatus := c.updateServiceDescriptor(logger, sd)
+		if errStatus != nil {
+			return retriableStatus, errStatus
+		}
+	}
+
+	// First filter to filter out the service descriptor if namespace is specified, to avoid
+	// treading on compositions trying to handleProcessResult and overwriting each other.
+	if c.namespace != "" {
+		nsServiceName, _ := deconstructNamespaceName(c.namespace)
+		if serviceName != nsServiceName {
+			logger.Sugar().Infof("Namespace %q can't possibly be created by SD %q. Skipping.", c.namespace, sd.Name)
+			return false, nil
+		}
+	}
+
+	filteredTransformer := &FilteredTransformer{
+		Transformer: c.sdTransformer,
+		Remove: func(info FormationObjectInfo) bool {
+			formLogger := logger.With(logz.NamespaceName(info.Namespace))
+			formLogger.Debug("Processing LocationDescriptor for ServiceDescriptor")
+			// filter the formationObjectDefList so we only handle the namespace we know
+			if info.Namespace != c.namespace && c.namespace != "" {
+				formLogger.Sugar().Infof("Namespace %q doesn't match namespace %q observed by the controller. Skipping formationObjectDef %v.",
+					info.Namespace, c.namespace, info.Location.Label)
+				return true
+			}
+			return false
+		},
+	}
+
+	formationObjectDefList, err := ProcessSD(sd, filteredTransformer)
+
+	foResults := make([]formationObjectResult, 0, len(formationObjectDefList))
+
+	if err != nil {
+		conflict, processRetryable, processErr := c.handleProcessResult(logger, serviceName, sd, foResults, false, false, err)
+		if conflict {
+			return false, nil
+		}
+
+		return processRetryable, processErr
+	}
+
+	var (
+		results        []*formationObjectResult
+		retryable      = false
+		deleteFinished = true
+		processFailure error
+	)
+
+	for _, result := range foResults {
+		formLogger := logger.With(logz.NamespaceName(result.namespace.Namespace))
+		formLogger.Debug("Processing LocationDescriptor for ServiceDescriptor")
+
+		var (
+			newResult        *formationObjectResult
+			processConflict  bool
+			processRetryable bool
+			processErr       error
+		)
+		if sd.ObjectMeta.DeletionTimestamp != nil {
```

Comment:
The original code treats updates and deletes as the same thing complicating the return result and flow control for that process. Here it's still pretty complicated. 

For example: what happens when the user removes a location?

Reply:
Yeah we don't handle deletions correctly at all.